File: js-pkg/sdk/src/connection.ts
import { HttpClient } from './http'
import { ClientToken } from './types'

export class YSweetConnection {
  private client: HttpClient

  constructor(clientToken: ClientToken) {
    this.client = new HttpClient(clientToken.url, clientToken.token ?? null)
  }

  /**
   * Returns an entire document, represented as a Yjs update byte string.
   *
   * This can be turned back into a Yjs document as follows:
   *
   * ```typescript
   * import * as Y from 'yjs'
   *
   * let update = await manager.getDocAsUpdate(docId)
   * let doc = new Y.Doc()
   * doc.transact(() => {
   *  Y.applyUpdate(doc, update)
   * })
   * ```
   *
   * @param docId
   * @returns
   */
  public async getDocAsUpdate(docId: string): Promise<Uint8Array> {
    const result = await this.client.request(`doc/${docId}/as-update`, 'GET')
    if (!result.ok) {
      throw new Error(`Failed to get doc ${docId}: ${result.status} ${result.statusText}`)
    }

    let buffer = await result.arrayBuffer()
    return new Uint8Array(buffer)
  }

  /**
   * Updates a document with the given Yjs update byte string.
   *
   * This can be generated from a Yjs document as follows:
   *
   * ```typescript
   * import * as Y from 'yjs'
   *
   * let doc = new Y.Doc()
   * // Modify the document...
   * let update = Y.encodeStateAsUpdate(doc)
   * await manager.updateDoc(docId, update)
   * ```
   *
   * @param docId
   * @param update
   */
  public async updateDoc(docId: string, update: Uint8Array): Promise<void> {
    const result = await this.client.request(`doc/${docId}/update`, 'POST', update)

    if (!result.ok) {
      throw new Error(`Failed to update doc ${docId}: ${result.status} ${result.statusText}`)
    }
  }
}

File: python/src/y_sweet_sdk/connection.py
import requests
from typing import Dict, Optional

from .update import UpdateContext


class DocConnection:
    def __init__(self, client_token: Dict[str, str]):
        self.base_url = client_token["baseUrl"].rstrip("/")
        self.token = client_token.get("token")
        self.doc_id = client_token["docId"]
        self.headers = {"Authorization": f"Bearer {self.token}"} if self.token else {}

    def _do_request(
        self, path: str, method: str = "GET", data: Optional[bytes] = None
    ) -> requests.Response:
        url = f"{self.base_url}/{path}"
        response = requests.request(method, url, headers=self.headers, data=data)
        response.raise_for_status()
        return response

    def get_as_update(self) -> bytes:
        """
        Returns an entire document, represented as a Yjs update byte string.

        Returns:
            bytes: The Yjs update as bytes.
        """
        response = self._do_request("as-update")
        return response.content

    def update_doc(self, update: bytes) -> None:
        """
        Updates a document with the given Yjs update byte string.

        Args:
            update (bytes): The Yjs update as bytes.
        """
        self._do_request("update", method="POST", data=update)

    def for_update(self) -> UpdateContext:
        return UpdateContext(self)
